src/python/pants/build_graph/address.py

```diff
@@ -172,15 +172,20 @@ def check_target_name(cls, spec_path: str, name: str) -> None:
                 ),
             )
 
-    def __init__(self, spec_path: str, target_name: str) -> None:
+    def __init__(
+        self, spec_path: str, target_name: str, *, generated_base_target_name: Optional[str] = None
```

Comment:
Open to suggestions on a better name than `generated_base_target_name`.

I do feel strongly about using the general term "generated subtargets" rather than file deps. File deps are part of this overall mechanism, but not the whole thing. Once we have the unification with file arguments, then it's possible to use generated subtargets without ever considering dependencies.

Reply:
One example where this framing would be very useful and that I plan to work on soon enough: when using [rsc](https://github.com/twitter/rsc) to parallelize scala compilations by generating subtargets, we might want to be able to generate many separate partitionings of the files within a scala target of various sizes, especially for the purpose of performance tuning. Especially if some or all of the non-rsc compilations performed over these partitions were to occur remotely via RBE or Scoot, I believe that testing @stuhood performed last year found that going all the way down to single-file parallelism did not necessarily achieve a speedup. So opening the door for arbitrary partitionings of targets is a good step, imho.

Reply:
Also, I think that `generated_base_target_name` is likely fine because of the clear and concise docstring.

Reply:
> So opening the door for arbitrary partitionings of targets is a good step, imho.

Agreed. Right now we do file-level to keep it simple, but that can easily be expanded to make more flexible for arbitrary partitioning.

Reply:
> then there would be file addresses without associated base targets on disk

That's part of the reason I shy away from approach #1. I'm very comfortable with this level of complexity that there is always an original owning target on disk that we can re-map to at any given moment.

Also, you'll see with the upcoming PR that generated subtargets result in _very_ little content in BUILD files. According to `wc -l`, our BUILD files in `src` and `test` are:

- before dep inference: 3371 lines
- after dep inference: 1450 lines
- after this feature is used: 855 lines

Reply:
> The advantage of having file-level dependencies (if you can get them, which we can in this case via inference) is that it lets you choose batches and partitions by coarsening the graph... which is easier than figuring out file-interdependencies if you don't have them already.

This makes perfect sense! Thanks!